### PR TITLE
Displaying the legend with specific dataset breaks the visual - fixed

### PR DESCRIPTION
**Pull request by zBritva**

### CODE REVIEW

File: src/visual.ts [hunk 1]
```diff
@@ -184,6 +184,9 @@ module powerbi.extensibility.visual {
         private static ShowAxisTitlePropertyName: string = "showAxisTitle";
         private static SecondShowAxisTitlePropertyName: string = "secShowAxisTitle";
 
+        private static SortDireacionDescending: string = "des";
```
Comment (ignatvilesov) [inline src/visual.ts]: SortDire`a`cionDescending. Please fix spelling

File: src/visual.ts [hunk 2]
```diff
@@ -184,6 +184,9 @@ module powerbi.extensibility.visual {
...
+        private static SortDireacionAscending: string = "asc";
```
Comment (ignatvilesov) [inline src/visual.ts]: Please fix spelling

File: src/visual.ts [hunk 3]
```diff
@@ -1477,7 +1480,7 @@ module powerbi.extensibility.visual {
                 legendData.dataPoints = [];
             }
 
-            let reducedLegends = [];
+            let reducedLegends: Array<IGrouppedLegendData> = [];
```
Comment (ignatvilesov) [inline src/visual.ts]: Please use short type declaration: `IGrouppedLegendData[]`

File: src/dataIntrefaces.ts [hunk 4]
```diff
@@ -75,6 +75,18 @@ module powerbi.extensibility.visual {
         seriesObjects: IDataViewObjects[][];
     }
 
+    export interface IGrouppedLegendData {
+        category: string;
+        index: number;
+        data: Array<any>;
```
Comment (ignatvilesov) [inline src/dataIntrefaces.ts]: Please use short type declaration: `any[]`. Why it it `any`?

File: src/visual.ts [hunk 5]
```diff
@@ -1557,13 +1560,17 @@ module powerbi.extensibility.visual {
...
-                    mekko.categoryLegends[index] = legend;
+                    mekko.categoryLegends[index] = <ILegend>legend;
                 }
             });
 
             legendParentsWithData.exit().remove();
             let svgHeight: number = 26;
             if (reducedLegends.length > 0) {
+                this.categoryLegends.forEach((legend, index) => {
+                    (<ILegendGroup>legend).position = +d3.select((<ILegendGroup>legend).element).style("top").replace("px", "");
```
Comment (ignatvilesov) [inline src/visual.ts]: Why do we need to remove `px`?

Comment (zBritva) [reply src/visual.ts]: @ignatvilesov for sorting as number

File: capabilities.json [hunk 6]
```diff
@@ -428,6 +428,40 @@
...
                 }
             }
+        },
+        "categoryColorStart": {
+            "displayName": "Categoty data colors start",
+            "displayNameKey": "Visual_CategoryDataColorsStart",
+            "properties": {
+                "categoryGradient": {
+                    "displayName": "Category gradient",
+                    "displayNameKey": "Visual_Enable",
```
Comment (ignatvilesov) [inline capabilities.json]: Please use the proper displayNameKey

File: capabilities.json [hunk 7]
```diff
@@ -428,6 +428,40 @@
...
+                }
+            }
+        },
+        "categoryColorEnd": {
+            "displayName": "Categoty data colors end",
+            "displayNameKey": "Visual_CategoryDataColorsEnd",
+            "properties": {
+                "categoryGradient": {
+                    "displayName": "Category gradient",
+                    "displayNameKey": "Visual_Enable",
```
Comment (ignatvilesov) [inline capabilities.json]: Please use the proper displayNameKey

File: package.json [hunk 8]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "powerbi-visuals-mekkochart",
-  "version": "1.9.2",
+  "version": "1.9.4",
```
Comment (ignatvilesov) [inline package.json]: I suppose you should use 1.9.3 as specified in pbiviz.json

File: src/columnChart/baseColumnChart.ts [hunk 9]
```diff
@@ -307,6 +307,8 @@ module powerbi.extensibility.visual.columnChart {
             // Allocate colors
             let legendAndSeriesInfo: LegendSeriesInfo = converterStrategy.getLegend(colors, defaultDataPointColor, "", dataPointSettings.categoryGradient, dataPointSettings.colorGradientEndColor.solid.color);
             let legend: MekkoLegendDataPoint[] = legendAndSeriesInfo.legend.dataPoints;
+            let currentColors: string = BaseColumnChart.serelizeColors(legend);
```
Comment (ignatvilesov) [inline src/columnChart/baseColumnChart.ts]: `serializeColors`? Is there the spelling mistake?

File: src/columnChart/baseColumnChart.ts [hunk 10]
```diff
@@ -376,6 +378,15 @@ module powerbi.extensibility.visual.columnChart {
             };
         }
 
+        private static serelizeColors(currentColors: MekkoLegendDataPoint[]): string {
```
Comment (ignatvilesov) [inline src/columnChart/baseColumnChart.ts]: `serializeColors`?

File: src/columnChart/baseColumnChart.ts [hunk 11]
```diff
@@ -1199,10 +1210,35 @@ module powerbi.extensibility.visual.columnChart {
...
                     break;
                 }
+                case "categoryColorEnd": {
+                    this.enumerateCategoryColors(enumeration, "categoryColorEnd", "End color");
+                    break;
+                }
+            }
+        }
+
+        private enumerateCategoryColors(instances: VisualObjectInstance[], objectName, label) {
```
Comment (ignatvilesov) [inline src/columnChart/baseColumnChart.ts]: It'd be nice to specify type of each argument

File: src/columnChart/baseColumnChart.ts [hunk 12]
```diff
@@ -1221,12 +1257,12 @@ module powerbi.extensibility.visual.columnChart {
             };
             instances[0].properties["enabled"] = this.data.sortlegend.enabled;
             instances[0].properties["direction"] = this.data.sortlegend.direction;
-            let allowedGroupByCategory = this.checkDataToFeatures();
+            // let allowedGroupByCategory = this.checkDataToFeatures();
```
Comment (ignatvilesov) [inline src/columnChart/baseColumnChart.ts]: Please remove commented code lines. They're useless.

File: src/columnChart/baseColumnChart.ts [hunk 13]
```diff
@@ -1315,9 +1351,9 @@ module powerbi.extensibility.visual.columnChart {
                 return;
             }
 
-            let allowedCategoryGradient = this.checkDataToFeatures();
+            // let allowedCategoryGradient = this.checkDataToFeatures();
 
-            if (allowedCategoryGradient) {
+            // if (allowedCategoryGradient) {
```
Comment (ignatvilesov) [inline src/columnChart/baseColumnChart.ts]: Please remove commented code lines. They're useless.

File: src/columnChart/baseColumnChart.ts [hunk 14]
```diff
@@ -1341,37 +1377,37 @@ module powerbi.extensibility.visual.columnChart {
...
+                //         instances.push({
+                //             objectName: "dataPoint",
+                //             displayName: "End color -" + category,
+                //             selector: ColorHelper.normalizeSelector((minValueDataPoint.identity as ISelectionId).getSelector(), true),
+                //             properties: {
+                //                 fill: { solid: { color: minValueDataPoint.color } }
+                //             },
+                //         });
+                //     });
+                // }
```
Comment (ignatvilesov) [inline src/columnChart/baseColumnChart.ts]: Please remove commented code lines. They're useless.

File: src/converterStrategy/baseConverterStrategy.ts [hunk 15]
```diff
@@ -81,78 +82,101 @@ module powerbi.extensibility.visual.converterStrategy {
...
-                    });
-                    categoryItemsCount[index] = mappedItems;
+            let restoredColors: any;
+            this.dataView.categories[categoryFieldIndex].values.forEach( (category: PrimitiveValue, index: number) => {
 
-                    if (colorGradient) {
-                        categoryItemsCount[index] = _.sortBy(categoryItemsCount[index], BaseConverterStrategy.SortField);
-                    }
+                const categorySelectionId: ISelectionId = this.visualHost.createSelectionIdBuilder()
+                .withCategory(this.dataView.categories[categoryFieldIndex], index)
```
Comment (ignatvilesov) [inline src/converterStrategy/baseConverterStrategy.ts]: Please format your files using VS Code

File: src/dataIntrefaces.ts [hunk 16]
```diff
@@ -75,6 +75,18 @@ module powerbi.extensibility.visual {
         seriesObjects: IDataViewObjects[][];
     }
 
+    export interface IGrouppedLegendData {
+        category: string;
+        index: number;
+        data: any[];
```
Comment (ignatvilesov) [inline src/dataIntrefaces.ts]: Can you use a particular type instead of any?

File: src/visual.ts [hunk 17]
```diff
@@ -1531,19 +1538,30 @@ module powerbi.extensibility.visual {
...
                 }
             }
 
+            let svgHeight: number = textMeasurementService.estimateSvgTextHeight({
+                fontFamily: MekkoChart.LegendBarTextFont,
+                fontSize: PixelConverter.toString(+legendProperties["fontSize"] + MekkoChart.LegendBarHeightMargin),
+                text: "AZ"
+            });
+
+            d3.select(this.rootElement.node()).selectAll("div.legendParent").remove();
```
Comment (ignatvilesov) [inline src/visual.ts]: Why does this removes DOM elements?

File: src/visual.ts [hunk 18]
```diff
@@ -1557,13 +1575,15 @@ module powerbi.extensibility.visual {
...
-                    mekko.categoryLegends[index] = legend;
+                    mekko.categoryLegends[index] = <ILegend>legend;
                 }
             });
 
-            legendParentsWithData.exit().remove();
-            let svgHeight: number = 26;
             if (reducedLegends.length > 0) {
+                this.categoryLegends.forEach((legend, index) => {
+                    (<ILegendGroup>legend).position = +d3.select((<ILegendGroup>legend).element).style("top").replace("px", "");
```
Comment (ignatvilesov) [inline src/visual.ts]: Legend actually implements method called getMargin that returns current offset. Can we use this method instead?

File: tsconfig.json [hunk 19]
```diff
@@ -13,7 +13,7 @@
     "node_modules/@types/d3/index.d.ts",
     "node_modules/@types/jquery/index.d.ts",
     "node_modules/@types/lodash/index.d.ts",
-    ".api/v1.6.0/PowerBI-visuals.d.ts",
+    ".api/v1.8.0/PowerBI-visuals.d.ts",
```
Comment (ignatvilesov) [inline tsconfig.json]: I think API version should be considered in _references.ts as well.

File: stringResources/en-US/resources.resjson [hunk 20]
```diff
@@ -45,5 +45,7 @@
 	"Visual_ColorDistibution": "Color distibution by value",
 	"Visual_ColorGradientEndColor": "End color of gradient",
 	"Visual_Default_Color": "Default color",
-	"Visual_Fill": "Fill"
+	"Visual_Fill": "Fill",
+  "Visual_GradientStartColor": "Start color",
+  "Visual_GradientEndColor": "End color"
```
Comment (ignatvilesov) [inline stringResources/en-US/resources.resjson]: Alignment

File: src/visual.ts [hunk 21]
```diff
@@ -1557,13 +1575,15 @@ module powerbi.extensibility.visual {
...
-                    mekko.categoryLegends[index] = legend;
+                    mekko.categoryLegends[index] = <ILegend>legend;
                 }
             });
 
-            legendParentsWithData.exit().remove();
-            let svgHeight: number = 26;
             if (reducedLegends.length > 0) {
+                this.categoryLegends.forEach((legend: ILegend, index: number) => {
+                    (<ILegendGroup>legend).position = +d3.select((<ILegendGroup>legend).element).style("top").replace("px", "");
```
Comment (ignatvilesov) [inline src/visual.ts]: Can't we use getMargin method of legend instance?

**Pull request merged**